src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.PendingBranchesCollection.cs

```diff
@@ -0,0 +1,150 @@
+                    }
+                }
+                else if (_labeledBranches is { } && _labeledBranches.TryGetValue(labelId.GetValueOrDefault(), out result))
+                {
+                    _labeledBranches.Remove(labelId.GetValueOrDefault());
+                }
+                else
+                {
+                    result = null;
+                }
+                return result;
+            }
+
+            internal void Add(int? labelId, PendingBranch branch)
+            {
+                if (labelId is null)
+                {
+                    _unlabeledBranches.Add(branch);
+                }
+                else
+                {
+                    var branches = GetOrAddBranchesToLabel(labelId.GetValueOrDefault());
+                    branches.Add(branch);
+                }
+            }
+
+            internal void AddRange(PendingBranchesCollection collection)
+            {
+                _unlabeledBranches.AddRange(collection._unlabeledBranches);
+                if (collection._labeledBranches is { })
+                {
+                    foreach (var pair in collection._labeledBranches)
+                    {
+                        var branches = GetOrAddBranchesToLabel(pair.Key);
+                        branches.AddRange(pair.Value);
+                    }
+                }
+            }
+
+            private ArrayBuilder<PendingBranch> GetOrAddBranchesToLabel(int labelId)
+            {
+                if (_labeledBranches is null)
+                {
+                    _labeledBranches = PooledDictionary<int, ArrayBuilder<PendingBranch>>.GetInstance();
+                }
+                if (!_labeledBranches.TryGetValue(labelId, out var branches))
+                {
+                    branches = ArrayBuilder<PendingBranch>.GetInstance();
+                    _labeledBranches.Add(labelId, branches);
+                }
+                return branches;
+            }
+
+            internal IEnumerable<PendingBranch> AsEnumerable()
+            {
+                return _labeledBranches is null ?
+                    _unlabeledBranches :
+                    asEnumerableCore();
+
+                IEnumerable<PendingBranch> asEnumerableCore()
+                {
+                    var labels = ArrayBuilder<int>.GetInstance();
+                    labels.AddRange(_labeledBranches.Keys);
+                    labels.Sort((x, y) => x - y);
+                    foreach (var branch in _unlabeledBranches)
+                    {
+                        yield return branch;
+                    }
+                    foreach (var label in labels)
+                    {
+                        var branches = _labeledBranches[label];
+                        foreach (var branch in branches)
+                        {
+                            yield return branch;
+                        }
+                    }
+                    labels.Free();
```

Comment:
This will not execute if the resulting `IEnumerable<PendingBranch>` is not fully exhausted. Essentially if there is a `break;` inside a `foreach`. Consider using a `try / finally` to ensure the `Free` is called in all cases.

Note: at the moment all of our `foreach` do fully walk the enumerable though.